feat: S3 CSV

sebastianswms opened the pull request:
tap-file with the ability to sync CSVs stored in Amazon S3 (and local files).

Functionality:
- Replaced sample-based schema generation with simple null/string schemas.
- A layer of abstraction for file-system creation allowing for easy additions.
- Support for both anonymous and credentialed S3 connections.
- RegEx-based file inclusion to allow selectivity with what is synced.

sebastianswms commented on README.md:
````diff
@@ -1,42 +1,31 @@
-tap-file --about --format=markdown
-```
--->
+| Setting                | Required | Default | Description |
````
The new version does mention batch config in the regen, but the description is blank:
```
| batch_config           | False    | None    |             |
```
Tried fiddling with it for a bit and couldn't get it work automatically, so I'll add it manually, along with the other auto-generated builtin settings.

sebastianswms commented on pyproject.toml:
```diff
@@ -11,17 +11,17 @@ keywords = [
 singer-sdk = { version="^0.27.0" }
-fs-s3fs = { version = "^1.1.1", optional = true }
+s3fs = ">=2023.5.0"
+fsspec = ">=2023.5.0"
```
Good to know, and ditto on s3fs.

sebastianswms commented on tap_file/tap.py:
```diff
@@ -14,31 +13,52 @@ class TapFile(Tap):
         ),
         th.Property(
-            "api_url",
+            "s3_access_key_secret",
```
I think that keeping separate config options for protocol, filepath, file_regex makes sense, instead of putting it all into a URI. That's a few extra config options, but:

- It's a one-time cost. No extra config options will be required as the number of supported sources increases.
- It's easier to implement and maintain on the backend, as each config option can be referenced individually rather than having to parse a URI,
- (opinion-based) It's a better experience for the user and is less likely to be the source of error.

As for `s3_anonymous_connection`, `s3_access_key`, and `s3_access_key_secret`, they could all be ditched if users are told to use the appropriate environment variables. I think that's probably the better long-term solution. It also has the additional advantage of easily allowing other S3 authentication methods supported by boto, such as security tokens, with no extra development work in the tap itself.

sebastianswms commented on meltano.yml:
```diff
@@ -1,29 +1,29 @@
+    - name: protocol
+    - name: filepath
+    - name: file_regex
+    - name: s3_anonymous_connection
```
Two potential situations:

- Access key and secret key are provided, but the connection should still be anonymous.
- Access key and secret key are _not_ provided, but the connection should still try to use them.

I don't think the first one would ever happen. If credentials are available, there's no downside to using them.

The second one could occur if you didn't want to provide config values in the tap but instead rely on boto to find the appropriate environment variables. However, that situation is sidestepped entirely if environment variables are _always_ used, as suggested in https://github.com/MeltanoLabs/tap-file/pull/3#discussion_r1219936285.

sebastianswms commented on meltano.yml:
```diff
@@ -1,29 +1,29 @@
+    - name: file_regex
+    - name: s3_anonymous_connection
+      kind: boolean
+    - name: s3_access_key
```
Batch config has been tested, both locally and with S3.

visch commented on README.md:
```diff
@@ -1,42 +1,79 @@
+| stream_map_config      | False    | None    | User-defined config values to be used within map expressions. |
+| flattening_enabled     | False    | None    | 'True' to enable schema flattening and automatically expand nested properties. |
+| flattening_max_depth   | False    | None    | The max depth to flatten schemas. | <!-- Manually added entries begin below. -->
+| batch_config           | False    | None    | Object containing batch configuration information. Has two child objects: `encoding` and `storage`. |
```
Can you link to the SDK docs here for this, it would probably be helpful 🤷

sebastianswms replied:
I've added a link to the Meltano documentation.

visch commented on pyproject.toml:
```diff
@@ -11,16 +11,21 @@ keywords = [
+
+[tool.poetry.extras]
+s3=["s3fs"]
+s3-batch=["fs-s3fs"]
```
Kind of interesting to make two seperate extras here I didn't think about that. What do you think about just shoving them both under the s3 extra? That's what the "standard" is for other taps today so it'd make sense for folks, and I think most of the overhead from the dependency is boto3 (and downstream from boto3) so the addition of fs-s3fs probably isn't a huge deal.

sebastianswms replied:
I've combined them into a single extra.

visch commented on pyproject.toml:
```diff
@@ -11,16 +11,21 @@ keywords = [
+singer-sdk = { version="^0.28.0" }
+s3fs = { version = "^2023.5.0", optional = true}
+fsspec = "^2023.5.0"
+ruff = "^0.0.270"
```
Linters we put under dev dependencies as they aren't needed to run the app. (black as well)

sebastianswms replied:
I've moved them to be dev dependencies.

visch commented on tap_file/client.py:
```diff
@@ -2,14 +2,136 @@
+                        target_options={"anon": True},
+                        cache_storage=cache_filepath,
+                    )
+                self.logger.warning(
```
Hmm, what is the benefit to having this be configurable? Does it save space? I think at a minimum we should just default to the OS's temp directory and call it a day. This hits on a concept called "convention over configuration" which is that we should recommend a conventional way of doing this, but if there is value we should allow folks to override the configuration.

I almost wonder if we shouldn't allow a config option here but :shrug:

sebastianswms replied:
My thinking was that for truly massive buckets, there wouldn't be space to cache everything on the local machine and it would be necessary to process each file individually. I'll switch the default to be the temp directory though, since that's probably a better "sensible default" than not using caching.

visch commented on tests/test_core.py:
```diff
@@ -1,22 +1,14 @@
-    "start_date": datetime.datetime.now(datetime.timezone.utc).strftime("%Y-%m-%d"),
-    # TODO: Initialize minimal tap config
-}
+SAMPLE_CONFIG = {}
```
We should get the basic tests running for this tap, we can setup an example S3 bucket with credentials (I can put them in the Repo, eventually we can migrate this to be Meltano's own S3 creds we'll just make a seperate issue and we can use AutoIDM's for now)

sebastianswms replied:
Will work on this after I complete tests for target-posgres.

visch reviewed (changes requested):
Looks really good just a couple questions / tweaks left

sebastianswms commented on meltano.yml:
```diff
@@ -1,29 +1,30 @@
-  - name: "tap-file"
-    namespace: "tap_file"
-    pip_url: -e .
+  - name: tap-file
```
I think putting everything under the banner of tap-file allows a lot of redundant code to be removed. If it were split into many similar taps, I think that would result in a lot of duplicated logic.

sebastianswms commented on tap_file/tap.py:
```diff
@@ -14,31 +15,67 @@ class TapFile(Tap):
+                "Ignored if `protocol!=s3`."
+            ),
+        ),
+        # Might want aliasing? Though aliases are also prepended with the name of the
```
Sounds good. I've removed the comment but will otherwise leave as-is.

visch commented:
When no items are matched we get this message with S3

```
2023-06-16 16:11:45,881 | WARNING  | tap-file             | Caching is not being used. The entire contents of each resource will be fetched during each read operation, which could be expensive.
Traceback (most recent call last):
  File "/home/visch/git/tap-file/.meltano/extractors/tap-file/venv/bin/tap-file", line 8, in <module>
    sys.exit(TapFile.cli())
  File "/home/visch/git/tap-file/.meltano/extractors/tap-file/venv/lib/python3.8/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/home/visch/git/tap-file/.meltano/extractors/tap-file/venv/lib/python3.8/site-packages/click/core.py", line 1054, in main
    with self.make_context(prog_name, args, **extra) as ctx:
  File "/home/visch/git/tap-file/.meltano/extractors/tap-file/venv/lib/python3.8/site-packages/click/core.py", line 920, in make_context
    self.parse_args(ctx, args)
  File "/home/visch/git/tap-file/.meltano/extractors/tap-file/venv/lib/python3.8/site-packages/click/core.py", line 1378, in parse_args
    value, args = param.handle_parse_result(ctx, opts, args)
  File "/home/visch/git/tap-file/.meltano/extractors/tap-file/venv/lib/python3.8/site-packages/click/core.py", line 2360, in handle_parse_result
    value = self.process_value(ctx, value)
  File "/home/visch/git/tap-file/.meltano/extractors/tap-file/venv/lib/python3.8/site-packages/click/core.py", line 2322, in process_value
    value = self.callback(ctx, self, value)
  File "/home/visch/git/tap-file/.meltano/extractors/tap-file/venv/lib/python3.8/site-packages/singer_sdk/tap_base.py", line 513, in cb_discover
    tap = cls(
  File "/home/visch/git/tap-file/.meltano/extractors/tap-file/venv/lib/python3.8/site-packages/singer_sdk/tap_base.py", line 104, in __init__
    self.mapper.register_raw_streams_from_catalog(self.catalog)
  File "/home/visch/git/tap-file/.meltano/extractors/tap-file/venv/lib/python3.8/site-packages/singer_sdk/tap_base.py", line 167, in catalog
    self._catalog = self.input_catalog or self._singer_catalog
  File "/home/visch/git/tap-file/.meltano/extractors/tap-file/venv/lib/python3.8/site-packages/singer_sdk/tap_base.py", line 311, in _singer_catalog
    for stream in self.streams.values()
  File "/home/visch/git/tap-file/.meltano/extractors/tap-file/venv/lib/python3.8/site-packages/singer_sdk/tap_base.py", line 129, in streams
    for stream in self.load_streams():
  File "/home/visch/git/tap-file/.meltano/extractors/tap-file/venv/lib/python3.8/site-packages/singer_sdk/tap_base.py", line 344, in load_streams
    for stream in self.discover_streams():
  File "/home/visch/git/tap-file/tap_file/tap.py", line 87, in discover_streams
    streams.CSVStream(self),
  File "/home/visch/git/tap-file/.meltano/extractors/tap-file/venv/lib/python3.8/site-packages/singer_sdk/streams/core.py", line 165, in __init__
    if not self.schema:
  File "/home/visch/.pyenv/versions/3.8.16/lib/python3.8/functools.py", line 967, in __get__
    val = self.func(instance)
  File "/home/visch/git/tap-file/tap_file/streams.py", line 44, in schema
    for field in reader.fieldnames:
TypeError: 'NoneType' object is not iterable
```

Maybe we could do better and throw an error with "Number of files matched is 0" try a new filepath / match (better wording needed)

visch commented:
If I put the data in a folder ie `/2023` I can't get the tap to pull the data. I've tried using `filepath:  derek-tap-filetesting/2023` and `file_regex: "2023/*"` but I didn't have any luck. We should allow subfolders to work in some way, we don't necessarily have to be able to auto traverse them. If this is hard then we should mention cleared in the Readme that we only work in root directories of an S3 bucket right now (which is fairly limited)

sebastianswms commented on .github/workflows/test.yml:
```diff
@@ -8,23 +7,31 @@ on: [push]
     - name: Test with pytest
       run: |
-        poetry run pytest
+        poetry run pytest --capture=no
```
I included it because `target-postgres` uses it.

sebastianswms commented on tap_file/tap.py:
```diff
@@ -14,31 +16,65 @@ class TapFile(Tap):
+            th.StringType,
             required=True,
-            description="Project IDs to replicate",
+            description="The path to obtain files from. Example: `/foo/bar`.",
```
This has a typo (`. ,`) but I'll make a similar change to add an example for how to specify an S3 bucket.

sebastianswms commented:
> Maybe we could do better and throw an error with "Number of files matched is 0"

Implemented.

> We should allow subfolders to work in some way

Implemented using the `filepath` setting.

visch commented:
Looks good to me I"m going to give it one more peak tomorrow morning with our tweaks we chatted about and a fresh sleep on it.
